[Tensorpipe Agent] Tracking Active Call Metrics

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38266 [Tensorpipe Agent] Add Call Counts to Metrics
* **#38265 [Tensorpipe Agent] Tracking Active Call Metrics**

Tracking the active calls counts in the TensorPipe Agent:
* clientActiveCalls: running count of un-responded/un-errored RPC's sent
* serverActiveCalls: running count of un-responded RPC's received
* serverAsyncCallCount: running count of received RPC's set to be completed asynchronously

This matches the metrics stored by the other 2 RPC Agents.

Differential Revision: [D21508957](https://our.internmc.facebook.com/intern/diff/D21508957/)

osalpekar committed:
[Tensorpipe Agent] Tracking Active Call Metrics

Tracking the active calls counts in the TensorPipe Agent:
* clientActiveCalls: running count of un-responded/un-errored RPC's sent
* serverActiveCalls: running count of un-responded RPC's received
* serverAsyncCallCount: running count of received RPC's set to be completed asynchronously

Differential Revision: [D21508957](https://our.internmc.facebook.com/intern/diff/D21508957/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Tracking Active Call Metrics"


Tracking the active calls counts in the TensorPipe Agent:
* clientActiveCalls: running count of un-responded/un-errored RPC's sent
* serverActiveCalls: running count of un-responded RPC's received
* serverAsyncCallCount: running count of received RPC's set to be completed asynchronously

This matches the metrics stored by the other 2 RPC Agents.

Differential Revision: [D21508957](https://our.internmc.facebook.com/intern/diff/D21508957/)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -312,6 +316,8 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
+
   // Don't need to hold lock while calling tensorpipe API.
   lock.unlock();
 
```
GitHub won't let me comment on lines you didn't touch, but further down there are two error-handling check in which I think we should also decrease `clientActiveCalls_`. Basically, every time we do `futureResponseMessage->setError()`.

lw replied:
I think there's still one occurrence you missed, on line 348.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -365,8 +371,10 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
               }
 
               threadPool_.run(
```
Unrelated, but is there a reason we defer this to the thread pool? Are we performing any blocking call?

osalpekar replied:
Yeah this seems like it shouldn't be deferred to the thread pool since we're just completing the future. cc @jiayisuse should we just do this in the main thread?

osalpekar committed:
Update on "[Tensorpipe Agent] Tracking Active Call Metrics"


Tracking the active calls counts in the TensorPipe Agent:
* clientActiveCalls: running count of un-responded/un-errored RPC's sent
* serverActiveCalls: running count of un-responded RPC's received
* serverAsyncCallCount: running count of received RPC's set to be completed asynchronously

This matches the metrics stored by the other 2 RPC Agents.

Differential Revision: [D21508957](https://our.internmc.facebook.com/intern/diff/D21508957/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Tracking Active Call Metrics"


Tracking the active calls counts in the TensorPipe Agent:
* clientActiveCalls: running count of un-responded/un-errored RPC's sent
* serverActiveCalls: running count of un-responded RPC's received
* serverAsyncCallCount: running count of received RPC's set to be completed asynchronously

This matches the metrics stored by the other 2 RPC Agents.

Differential Revision: [D21508957](https://our.internmc.facebook.com/intern/diff/D21508957/)

[ghstack-poisoned]
